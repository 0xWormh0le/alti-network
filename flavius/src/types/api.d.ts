type QueryParamValue = string | number | string[] | number[] | null | undefined | boolean
type QueryParams = { [k: string]: QueryParamValue }

type ApiClientReturn<TState extends any, TCallerArgs extends any> = [
  Maybe<TState>,
  Maybe<{ error: string }>,
  boolean,
  ApiCaller<TCallerArgs>
]

type ApiErrorCallback = (err: any) => void
type ApiSuccessCallback = (response: any) => void

interface ApiClientInit {
  defaultPageSize?: number
  pageCountCacheTtl?: number
  handleError?: ApiErrorCallback
  handleSuccess?: ApiSuccessCallback
  immediateTrigger?: boolean
}

interface ApiCaller<T> {
  call: (args?: T) => void
}

interface PaginationArgs {
  pageSize: number
  pageCountCacheTtl: number
  pageCountLastUpdated: number
}

interface MetaArgs<T, TParams = unknown> {
  handleError?: (err: Error) => void
  handleSuccess?: (rs: T) => void
  params?: (QueryParams & TParams) | undefined
  preventFirstCall?: boolean
  triggeringKeys?: string[]
}

interface InternalMetaArgs<T, TParams = unknown> extends MetaArgs<T, TParams> {
  trigger: boolean
  resetTrigger: () => void
  handleError: (err: Error) => void
  handleSuccess: (rs: T) => void
  params: (QueryParams & TParams) | undefined
}
interface ApiCallArgs<T, TArgs> {
  requestDetails?: TArgs
  meta?: MetaArgs<T>
}

type WithPagination<T> = T & {
  pageNumber: number
  pageSize: number
}
