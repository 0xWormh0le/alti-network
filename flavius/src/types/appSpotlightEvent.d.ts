// Display interfaces for each kind of event types are represented below

// event type: people
interface DisplayAppSpotlightPeople {
  avatar: string
  name: string
  email: string
}

// event type: downloads
interface DisplayAppSpotlightEventDownloads {
  view: IFile
  file: IFile
  owner: IPerson
  datetime: number
}

interface BaseAppSpotlightEvent {
  targetPeople: IPerson[]
  files: IFile[]
  eventName: string
  eventDescription: string
  severity: SeverityRange
  datetime: number
}

interface AppSpotlightEvent extends BaseAppSpotlightEvent {
  eventId: string
}

// event type: risks
interface DisplayAppSpotlightRisk extends BaseRisk {
  view: FileResponseWithRisk
  file: FileCellWithRisk
  description: DisplayRiskDescription
  creator: IPerson
  owner: FileOwner
  action: RiskAction
}
