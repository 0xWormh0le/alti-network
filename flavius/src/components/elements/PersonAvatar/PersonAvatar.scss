.PersonAvatar {
  display: flex;
  align-items: center;

  &__clickable {
    cursor: pointer;
  }
}
