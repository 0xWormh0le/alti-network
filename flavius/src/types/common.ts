export enum UsageKind {
  personal = 'personal',
  work = 'work'
}

export enum UserKind {
  user = 'user',
  person = 'person',
  serviceAccount = 'serviceAccount'
}

export enum AccessLevel {
  member = 'member',
  manager = 'manager',
  owner = 'owner',
  admin = 'admin'
}

export enum RiskCategoryType {
  SHARING_RISKS = 'Sharings',
  RELATIONSHIP_RISKS = 'Relationships',
  ACTIVITY_RISKS = 'Actions',
  INFORMATIONAL_RISKS = 'Informatives'
}

export enum FileActivityType {
  ACCESS_GRANTED = 'ACCESS_GRANTED',
  CHANGE_FOLDER_PERMISSION = 'CHANGE_FOLDER_PERMISSION',
  CONTENT_ACCESS = 'CONTENT_ACCESS',
  DOWNLOAD = 'DOWNLOAD',
  FILE_WATERMARKED_DOWNLOAD = 'FILE_WATERMARKED_DOWNLOAD',
  ITEM_DOWNLOAD = 'ITEM_DOWNLOAD',
  ITEM_PREVIEW = 'ITEM_PREVIEW',
  ITEM_SHARED = 'ITEM_SHARED',
  ITEM_SHARED_CREATE = 'ITEM_SHARED_CREATE',
  ITEM_SHARED_UPDATE = 'ITEM_SHARED_UPDATE',
  ITEM_SYNC = 'ITEM_SYNC',
  SHARE = 'SHARE'
}

export enum PermissionRole {
  own = 'own',
  read = 'read',
  write = 'write'
}

export enum PermissionShared {
  external = 'external',
  internal = 'internal',
  none = 'none'
}

export enum PermissionType {
  anyone = 'anyone',
  domain = 'domain',
  user = 'user'
}

export enum PermissionItemStatus {
  active = 'ACTIVE',
  pending = 'PENDING',
  removed = 'REMOVED',
  cannotBeRemoved = 'CANNOT_BE_REMOVED'
}

export enum SensitiveContentStatus {
  DETECTION_PENDING = 'pending',
  FOUND_SENSITIVE_CONTENT = 'yes',
  FOUND_NO_SENSITIVE_CONTENT = 'no',
  FAILED_DETECTION = 'failed'
}

export enum FileGridNavType {
  TIMELINE = 'timeline',
  INTERNAL = 'internal',
  EXTERNAL = 'external',
  SENSITIVE = 'sensitive',
  FILES_IN_FOLDER = 'filesInFolder',
  DETAILS = 'details'
}
