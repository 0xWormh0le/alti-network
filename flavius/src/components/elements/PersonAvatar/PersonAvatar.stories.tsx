import React from 'react'
import { Story, Meta } from '@storybook/react/types-6-0'
import PersonAvatar, { PersonAvatarProps } from './PersonAvatar'

export default {
  title: 'Elements/PersonAvatar',
  component: PersonAvatar
} as Meta

const Template: Story<PersonAvatarProps> = (args) => <PersonAvatar {...args} />

export const Default = Template.bind({})
Default.args = {
  person: {
    name: {
      givenName: 'Bobbie',
      familyName: 'Name',
      fullName: 'Bobbie Name'
    },
    accessCount: 1,
    avatar: {
      url: '',
      url_etag: ''
    },
    displayName: 'Bobbie'
  } as any
}
