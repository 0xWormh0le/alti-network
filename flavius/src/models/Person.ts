import { UserKind, AccessLevel } from 'types/common'
import UI_STRINGS from 'util/ui-strings'
import anonymousUserIcon from 'icons/anonymous-user.svg'
class Person implements IPerson {
  public accessLevel: AccessLevel
  public name: Name
  public projectId: string
  public primaryEmail: Email
  public userKind: UserKind
  public altnetId?: string
  public recoveryEmail?: Email
  public phones: Phone[]
  public externalIds: UserId[]
  public avatar?: Avatar
  public internal?: boolean // deprecated thanks to userKind
  public anonymous?: boolean
  public internalCount?: number
  public externalCount?: number
  public emails: Email[]
  public accessCount?: number
  public riskCount?: number
  public orgUnitPath?: string
  public isEnrolledInMFA?: boolean
  public etag?: string
  public creationTime?: number
  public lastLoginTime?: number
  public lastModifiedTime?: number
  public notes?: Notes
  public lastRemovedPermissions?: number | null
  public title?: string
  public department?: string
  public phoneNumber?: string

  constructor(person: IPerson) {
    this.altnetId = person.altnetId
    this.projectId = person.projectId
    this.userKind = person.userKind
    this.accessLevel = person.accessLevel
    this.name = person.name || {
      givenName: '',
      familyName: '',
      fullName: ''
    }
    this.primaryEmail = person.primaryEmail
    this.recoveryEmail = person.recoveryEmail
    this.phones = person.phones || []
    this.externalIds = person.externalIds || []
    this.orgUnitPath = person.orgUnitPath
    this.etag = person.etag
    this.isEnrolledInMFA = person.isEnrolledInMFA
    this.creationTime = person.creationTime
    this.lastLoginTime = person.lastLoginTime
    this.lastModifiedTime = person.lastModifiedTime
    this.notes = person.notes
    this.avatar =
      this.name.fullName === UI_STRINGS.RISKS.ANONYMOUS_USER ? { url: anonymousUserIcon, url_etag: '' } : person.avatar
    this.title = person.title
    this.department = person.department
    this.phoneNumber = person.phoneNumber
    this.internal = person.internal
    this.anonymous = !person.primaryEmail?.address || person.primaryEmail?.address?.indexOf('@') === -1 // altnetId: person.personId?.indexOf('@') === -1 &&
    this.internalCount = person.internalCount
    this.externalCount = person.externalCount
    this.emails = person.emails || []
    this.accessCount = person.accessCount
    this.riskCount = person.riskCount
    this.lastRemovedPermissions = person.lastRemovedPermissions
  }

  public get displayName(): string {
    if (this.name.givenName && this.name.familyName) {
      return `${this.name.givenName} ${this.name.familyName}`
    } else if (this.anonymous) {
      return UI_STRINGS.RISKS.ANONYMOUS_USER
    } else if (this.primaryEmail?.address) {
      const emailFirst = this.primaryEmail?.address.split('@')
      return emailFirst && emailFirst.length > 0 ? emailFirst[0] : this.primaryEmail?.address
    } else {
      return this.primaryEmail?.address ?? '' // altnetId:
    }
  }

  /*
   * For transform value from Map to Object in griddle-react custom components, e.g. SimplePersonCell, ActorCell
   * This is a workaround solution, once we refactor the table to NOT use griddle-react, this method shall be gone
   */
  static fromMap(personMap: { get: (x: keyof IPerson) => any }): Person {
    return new Person({
      primaryEmail: personMap.get('primaryEmail') && (Object.fromEntries(personMap.get('primaryEmail')) as Email),
      recoveryEmail: personMap.get('recoveryEmail') && (Object.fromEntries(personMap.get('recoveryEmail')) as Email),
      avatar: personMap.get('avatar') && (Object.fromEntries(personMap.get('avatar')) as Avatar),
      name: personMap.get('name') && (Object.fromEntries(personMap.get('name')) as Name),
      internal: personMap.get('internal'),
      internalCount: personMap.get('internalCount'),
      externalCount: personMap.get('externalCount'),
      emails:
        personMap.get('emails') &&
        (Array.from(personMap.get('emails')).map((email: any) => Object.fromEntries(email)) as Email[]),
      accessCount: personMap.get('accessCount'),
      riskCount: personMap.get('riskCount'),
      altnetId: personMap.get('altnetId'),
      projectId: personMap.get('projectId'),
      accessLevel: personMap.get('accessLevel'),
      userKind: personMap.get('userKind'),
      phones:
        personMap.get('phones') &&
        (Array.from(personMap.get('phones')).map((email: any) => Object.fromEntries(email)) as Phone[]),
      externalIds:
        personMap.get('externalIds') &&
        (Array.from(personMap.get('externalIds')).map((email: any) => Object.fromEntries(email)) as UserId[]),
      orgUnitPath: personMap.get('orgUnitPath'),
      etag: personMap.get('etag'),
      isEnrolledInMFA: personMap.get('isEnrolledInMFA'),
      creationTime: personMap.get('creationTime'),
      lastLoginTime: personMap.get('lastLoginTime'),
      lastModifiedTime: personMap.get('lastModifiedTime'),
      notes: personMap.get('notes') && (Object.fromEntries(personMap.get('notes')) as Notes)
    })
  }
}

export default Person
