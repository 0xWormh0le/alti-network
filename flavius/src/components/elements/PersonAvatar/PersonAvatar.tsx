import React, { CSSProperties } from 'react'
import { Link } from 'react-router-dom'
import cx from 'classnames'
import { modalBasePath, searchWithoutModalParams } from 'util/helpers'
import Person from 'models/Person'
import { BaseAvatar } from '@altitudenetworks/component-library/'
import Tooltip from 'components/widgets/Tooltip'
import CONSTANTS from 'util/constants'
import UI_STRINGS from 'util/ui-strings'
import './PersonAvatar.scss'

export interface PersonAvatarProps {
  person: Person | LitePerson
  className?: string
  clickable?: boolean
  style?: CSSProperties
}

const PersonAvatar: React.FC<PersonAvatarProps> = ({ person, className, clickable = false, style = {} }) => {
  const primaryTooltipText =
    person.displayName !== UI_STRINGS.RISKS.ANONYMOUS_USER
      ? person.displayName
        ? person.displayName
        : person.name.fullName
      : ''

  const secondaryTooltipText = person.displayName === person.primaryEmail?.address ? '' : person.primaryEmail?.address
  // altnetId: person.primaryEmail?.address
  const baseAvatarWrapper = (
    <Tooltip
      text={primaryTooltipText}
      secondaryText={secondaryTooltipText}
      id={`tooltip-avatar-${person.primaryEmail?.address}-id`}>
      <span className={cx('PersonAvatar', className, { PersonAvatar__clickable: clickable })}>
        <BaseAvatar
          style={{ ...style }}
          src={person.avatar}
          name={person.displayName || person.name.fullName}
          colorList={CONSTANTS.COLOR_LIST}
        />
      </span>
    </Tooltip>
  )
  // altnetId: person.primaryEmail?.address
  return clickable ? (
    <Link
      to={`${modalBasePath()}/spotlight/${encodeURIComponent(
        person.primaryEmail?.address ?? ''
      )}${searchWithoutModalParams()}`}>
      {baseAvatarWrapper}
    </Link>
  ) : (
    baseAvatarWrapper
  )
}

export default PersonAvatar
