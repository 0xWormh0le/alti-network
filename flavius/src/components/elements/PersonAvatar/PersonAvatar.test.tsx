import React from 'react'
import { renderWithRouter } from 'test/support/helpers'
import PersonAvatar from './PersonAvatar'
import Person from 'models/Person'

import { UsageKind, UserKind, AccessLevel } from 'types/common'

const person = new Person({
  name: {
    givenName: 'Rory',
    familyName: 'Garrett',
    fullName: 'Rory Garrett'
  },
  primaryEmail: {
    address: 'rorygarret@mycompany.org',
    kind: UsageKind.personal,
    primary: true,
    accessCount: 0,
    riskCount: 0,
    lastDeletedPermissions: 0
  },
  recoveryEmail: {
    address: 'test@email.com',
    kind: UsageKind.personal,
    primary: true,
    accessCount: 0,
    riskCount: 0,
    lastDeletedPermissions: 0
  },
  avatar: {
    url: 'avatar',
    url_etag: 'avatar'
  },
  internalCount: 1,
  externalCount: 0,
  emails: [],
  internal: true,
  accessCount: 0,
  riskCount: 99,
  altnetId: 'rorygarret@mycompany.org',
  projectId: '',
  accessLevel: AccessLevel.member,
  userKind: UserKind.user,
  phones: [],
  externalIds: [],
  orgUnitPath: '',
  etag: '',
  isEnrolledInMFA: false,
  creationTime: 0,
  lastLoginTime: 0,
  lastModifiedTime: 0,
  notes: {
    content: '',
    content_type: ''
  }
})

describe('PersonAvatar', () => {
  describe('when the `clickable` prop is false', () => {
    it('does not render any links and PersonAvatar__clickable classname is not included', () => {
      const { container } = renderWithRouter(<PersonAvatar person={person} clickable={false} />)
      expect(container.firstChild).toMatchSnapshot()
    })
  })

  describe('when the `clickable` prop is true', () => {
    it('wraps the element with a link (anchor tag) to the Spotlight page using the id from the person', () => {
      const { container } = renderWithRouter(<PersonAvatar person={person} clickable={true} />)
      expect(container.firstChild).toMatchSnapshot()
    })
  })

  it('renders tooltip for avatar', () => {
    const { container } = renderWithRouter(<PersonAvatar person={person} clickable={true} />)
    expect(container.querySelector('.PersonAvatar')).toHaveClass('trigger')
  })
})
