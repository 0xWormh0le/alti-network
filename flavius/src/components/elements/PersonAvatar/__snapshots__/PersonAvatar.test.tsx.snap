// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PersonAvatar when the \`clickable\` prop is false does not render any links and PersonAvatar__clickable classname is not included 1`] = `
<span
  class="PersonAvatar trigger"
>
  <div
    class="BaseAvatar "
  >
    <div
      class="BaseAvatar__image"
    >
      <img
        alt="RG"
        src="avatar"
        style="width: 2em;"
      />
    </div>
  </div>
</span>
`;

exports[`PersonAvatar when the \`clickable\` prop is true wraps the element with a link (anchor tag) to the Spotlight page using the id from the person 1`] = `
<a
  href="/dashboard/spotlight/rorygarret%40mycompany.org"
>
  <span
    class="PersonAvatar PersonAvatar__clickable trigger"
  >
    <div
      class="BaseAvatar "
    >
      <div
        class="BaseAvatar__image"
      >
        <img
          alt="RG"
          src="avatar"
          style="width: 2em;"
        />
      </div>
    </div>
  </span>
</a>
`;
